### PR TITLE
CLOUDSTACK-10381: Fix password reset / reset ssh key with ConfigDrive

### PR DESCRIPTION
**Pull request by fmaximus**
## Description

Let ConfigDriveElement remember the new password, 
so it can be put on the ConfigDrive iso while vm is starting.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [ ] I have added tests to cover my changes.
- [ ] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by PaulAngus** [comment]
From what i read, it looks to me like you're using the standard encryption algorithm to store the users' password - I would be quite strongly opposed to storing these passwords (and i think that people would have compliance issues) outside of a proper 'vault'.

I'd also ask if the password is removed from the config drive once a VM has booted.

File: server/src/com/cloud/network/element/ConfigDriveNetworkElement.java [hunk 1]
```diff
@@ -237,6 +252,20 @@ public boolean saveUserData(final Network network, final NicProfile nic, final V
         return canHandle(network.getTrafficType());
     }
 
+    /**
+     * Store password in vm details so it can be picked up during VM start.
+     */
+    private void storePasswordInVmDetails(VirtualMachineProfile vm) {
```
Comment (rafaelweingartner) [inline server/src/com/cloud/network/element/ConfigDriveNetworkElement.java]: Would you mind creating a unit test for this method?

File: server/src/com/cloud/network/element/ConfigDriveNetworkElement.java [hunk 2]
```diff
@@ -223,7 +231,14 @@ public boolean saveSSHKey(final Network network, final NicProfile nic, final Vir
         if (vm != null && vm.getVirtualMachine().getState().equals(VirtualMachine.State.Running)) {
             throw new CloudRuntimeException("VM should to stopped to reset password");
         }
-        return canHandle(network.getTrafficType());
+
+        final boolean canHandle = canHandle(network.getTrafficType());
+
+        if (canHandle) {
+            storePasswordInVmDetails(vm);
```
[the remark this answers is not in the document]
Comment (rhtyd) [reply server/src/com/cloud/network/element/ConfigDriveNetworkElement.java]: I may be wrong, but when I tested it I could see different passwords in the isos before shutdown and post-reset password and startup of the VM @fmaximus. Have you tested this against a normal config drive enabled network such as a L2 network?

[the remark this answers is not in the document]
Comment (fmaximus) [reply server/src/com/cloud/network/element/ConfigDriveNetworkElement.java]: SaveSshKey implicitly also resets the password. So I'm also saving it in this case.

**Comment by fmaximus** [comment]
I've used the same code as is used in VRElement, when VR is stopped.
There it's delayed to when the VM is starting, as by then the VR is running.

We've looked at removing the password from the config drive, it soon becomes complex.
- As you can't be certain that the password has been requested (like in VR), 
  you don't know for sure when to remove it. 
- The only way to remove it is to build a new iso, and change the config drive disk.
  This can only be done safely when the VM is stopped.

**Comment by rhtyd** [comment]
@fmaximus yes, it is a requirement that VM is stopped for doing password reset and this also ensures that a new ISO will be built when VM is started (if config drive is enabled/available). Per the apidocs, similar to the reset SSH key API I've added a check that will not allow reset password API to work if VM is not stopped. /cc @PaulAngus @DaanHoogland

**Comment by PaulAngus** [comment]
LGTM - tested by;
creating an L2 network with configdrive
marking a template as password enabled and then starting the built-in template
inspecting config drive i can find password file with correct password inside
stopping and starting vm, password file is removed
stopping vm and resetting password
after starting VM password file has correct password
stopping and starting vm, password file is removed again.

**Comment by DaanHoogland** [comment]
@rhtyd you approve and say LGTM but then say "if..." effectively -1ing the PR. the password will remain available in the configdrive until it is rereated, won't it? @fmaximus am I seeing this wrong?

**Pull request merged**